### ISSUE TITLE
Linux support

### ISSUE THREAD

**Issue by krymtkts** [krymtkts/pocof#85]
Currently, all tests pass in non-interactive mode.
However, it is known that the following error occurs in interactive mode.

```
PS /src> Get-ChildItem | Select-Pocof
Select-Pocof: The method or operation is not implemented.
```

detailed information.

```
PS /src> $ErrorView = 'DetailedView'
PS /src> $Error

Exception             :
    Type       : System.NotImplementedException
    TargetSite :
        Name          : GetBufferContents
        DeclaringType : Microsoft.PowerShell.ConsoleHostRawUserInterface, Microsoft.PowerShell.ConsoleHost, Version=7.3.10.500, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : Microsoft.PowerShell.ConsoleHost.dll
    Message    : The method or operation is not implemented.
    Source     : Microsoft.PowerShell.ConsoleHost
    HResult    : -2147467263
    StackTrace :
   at Microsoft.PowerShell.ConsoleHostRawUserInterface.GetBufferContents(Rectangle rectangle)
   at System.Management.Automation.Internal.Host.InternalHostRawUserInterface.GetBufferContents(Rectangle r)
   at pocof.PocofScreen.RawUI..ctor(PSHostRawUserInterface rui) in /src/src/pocof/UI.fs:line 55
   at pocof.PocofScreen.init(PSHostRawUserInterface rui, String prompt, FSharpFunc`2 invoke) in /src/src/pocof/UI.fs:line 173
   at pocof.SelectPocofCommand.interact(InternalConfig conf, InternalState state, Position pos, PSHostRawUserInterface rui, FSharpFunc`2 invoke) in /src/src/pocof/Library.fs:line 44
   at pocof.SelectPocofCommand.EndProcessing() in /src/src/pocof/Library.fs:line 191
CategoryInfo          : NotImplemented: (:) [Select-Pocof], NotImplementedException
FullyQualifiedErrorId : NotImplementedException,pocof.SelectPocofCommand
InvocationInfo        :
    MyCommand        : Select-Pocof
    ScriptLineNumber : 1
    OffsetInLine     : 17
    HistoryId        : 2
    Line             : Get-ChildItem | pocof
    PositionMessage  : At line:1 char:17
                       + Get-ChildItem | pocof
                       +                 ~~~~~
    InvocationName   : pocof
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo :
```

[Unix system executables terminate with error if transcription enabled · Issue #1920 · PowerShell/PowerShell](https://github.com/PowerShell/PowerShell/issues/1920)

Is this related this issue?

**Comment by krymtkts** [comment]
Implemented workaround for `NotImplementedException` raised by `Microsoft.PowerShell.ConsoleHostRawUserInterface.GetBufferContents`. #86

**Comment by krymtkts** [comment]
[PS-GuiCompletion does not work in PowerShell Core 7.0 on Linux · Issue #13 · cspotcode/PS-GuiCompletion](https://github.com/cspotcode/PS-GuiCompletion/issues/13#issuecomment-620084134)

> Similar functionality, such as the Menucomplete feature of PSReadLine, work around this limitation by writing several blank lines to the terminal, scolling everything upward and creating an empty region at the bottom of the screen. They draw their "gui" into this empty region, knowing they can erase it entirely when they're done.

**Comment by krymtkts** [comment]
PoC script using PSReadLine style buffer control.

```powershell
# test script for interactive console window.
# currently cannot prevent mouse scrolling.
function Test-ConsoleWindowWithoutBufferCleaning {
    # backup cursor x position.
    $x = [Console]::CursorLeft

    # add lines to the end of the screen for scrolling using the PSReadLine method.
    # https://github.com/PowerShell/PSReadLine/blob/e57f7d691d8df8c1121fddf47084f96aea74a688/PSReadLine/DisplayBlockBase.cs#L17-L24
    $h = [Console]::WindowHeight
    0..($h - 1) | ForEach-Object { [Console]::Write("`n") }

    # write contents.
    $yy = [COnsole]::CursorTop
    $yy .. 0 | ForEach-Object {
        [Console]::SetCursorPosition($x, $_)
        [Console]::Write("Line: $_")
    }

    # some interaction.
    $arr = @()
    while ($true) {
        $k = [Console]::ReadKey()
        $arr = @($k) + $arr
        if ($k.Key -eq 'Enter') {
            break
        }
        else {
            $i = 0
            (($arr | Out-String ) -split "`n" ) | Select-Object -First $h | ForEach-Object {
                [Console]::SetCursorPosition(0, $i)
                [Console]::Write($_) 
                $i += 1
            }
        }
    }

    # clear contests.
    $w = [Console]::WindowWidth
    0..($h - 1) | ForEach-Object {
        [Console]::SetCursorPosition(0, $_)
        [Console]::Write(' ' * $w)
    }

    # restore cursor position.
    [Console]::SetCursorPosition($x, 0)
}

Test-ConsoleWindowWithoutBufferCleaning
```

but cannot prevent mouse scrolling.

![poc](https://github.com/krymtkts/pocof/assets/13184040/5af6630d-5fb2-482d-b25d-fa415d56765c)

**Comment by krymtkts** [comment]
PoC script using PSReadLine style buffer control written in F#.

```fsharp
open System

[<TailCall>]
let rec readAndDisplay h w arr =
    let k = Console.ReadKey(true)
    let arr = k :: arr

    if k.Key = ConsoleKey.Enter then
        ()
    else
        let take =
            match List.length arr < h with
            | true -> List.length arr
            | _ -> h

        arr
        |> List.take take
        |> List.map (fun k -> $"[{k.Key} - {k.Modifiers}]")
        |> List.iteri (fun i s ->
            Console.SetCursorPosition(0, i)
            let s = $"{i} {s}"

            Console.Write(s + String.replicate (w - String.length s) " "))

        readAndDisplay h w arr

// test script for interactive console window.
// currently cannot prevent mouse scrolling.
let testConsoleWindowWithoutBufferCleaning () =
    // backup cursor x position.
    let x = Console.CursorLeft

    // add lines to the end of the screen for scrolling using the PSReadLine method.
    let h = Console.WindowHeight
    let w = Console.WindowWidth

    String.replicate (h - 1) "\n" |> Console.Write

    // write contents.
    let yy = Console.CursorTop

    Console.SetCursorPosition(0, 0)

    [ 0..yy ]
    |> List.map (fun i ->
        let s = $"Line: %d{i}"
        s + String.replicate (w - s.Length) " ")
    |> String.concat "\n"
    |> Console.Write
    // read and display.
    readAndDisplay h w []

    // clear contests.
    [ 0 .. (h - 1) ]
    |> List.iter (fun i ->
        Console.SetCursorPosition(0, i)
        Console.Write(String.replicate w " "))

    // restore cursor position.
    Console.SetCursorPosition(x, 0)

testConsoleWindowWithoutBufferCleaning ()
```

![poc](https://github.com/krymtkts/pocof/assets/13184040/c8822f0a-fd13-4083-9a62-ac9c8dc56d58)

**Comment by krymtkts** [comment]
pocof buffer controle refined in #124

**Comment by krymtkts** [comment]
This issue is planned to be closed in the 0.9.0 release.

**Comment by krymtkts** [comment]
close by #128
